### ISSUE TITLE
Analyse: Nominatim hierarchy

### ISSUE THREAD

**Issue by klokan** [geometalab/OSMNames#5]
Very important point on which we need to know soon is how exactly Nominatim builds the "display_name" variable - and what part of the hierarchy reconstruction is pre-calculated (and stored in SQL tables) during the import of OSM planet and what part is made during query time.

Purpose of this ticket is to document and link relevant code related to gazetteer hierarchy made from OSM in Nominatim.

**Comment by and01** [comment]
**display_name**

excerpt from _search-batch-json.php_

`$aPlace['display_name'] = $aPointDetails['name'];
`

The name appears in the search results and is fetched from the _placex_ table. _name_ in _placex_ is an hstore and could look like this:

` "name"=>"Eschweiler", "name:lb"=>"Eschweiler" `

**Comment by klokan** [comment]
display_name in sample query "Paris" contains in Nominatim:
"Paris, Ile-de-France, Metropolitan France, France"
see: http://nominatim.klokantech.com/?q=paris&format=jsonv2&addressdetails=1

**Comment by and01** [comment]
It looks like _get_addressdata(in_place_id)_ in _functions.sql _provides the desired results which are later on filtered by language and somewhat sorted. I will delve into that later on..

**Comment by klokan** [comment]
This should be done until next week (new milestone and meeting agreed).

- Mazen city extracts: "Zurich" at https://mapzen.com/data/metro-extracts/ as sample data in local Nominatim installation.

- Essential:
  - How to Nominatim links geometry to place nodes (Dorf/Gemeinde)?
  - Hierarchy (belongs to Switzerland, belongs to Kanton, etc) reconstruction
  - The hierarchy in "display_name" discussed above in details
  - Construction of geometry if relations are missing (no direct link in data) and if relations are present such as in case of:

http://nominatim.openstreetmap.org/search.php?q=Zurich&polygon=1&viewbox=
http://nominatim.openstreetmap.org/details.php?place_id=144801142
with `boundary:administrative` http://www.openstreetmap.org/relation/1682248
and `place:city` node "as label" http://www.openstreetmap.org/node/240025182

**Comment by manuelroth** [comment]
- Custom extracts can be made with this tool http://extract.bbbike.org/
- You create an extract job and after 5-10 Minutes the extract is ready to download.
- On the [status page](http://download.bbbike.org/osm/extract/) you can check if your job is already done

**Comment by and01** [comment]
**Here is a more detailled description of how nominatim gets the display_name**

Example for search query 'Helvetiaplatz' on http://localhost:8080/?q=helvetiaplatz&format=jsonv2&addressdetails=1

When executing a search query the following steps are taken:

1. nominatim tries to find place_ids in placex relevant to the search
2. with the place_id s found, the SQL query get_address_by_language delivers the results for the display_name variable

**1. search word**

`select word_id,word_token, word, class, type, country_code, operator, search_name_count from word where word_token in (' helvetia pl','helvetia pl',' helvetia','helvetia',' pl','pl')`
```
 word_id |  word_token  |     word      | class | type | country_code | operator | search_name_count 
---------+--------------+---------------+-------+------+--------------+----------+-------------------
      15 |  pl          | plaza         |       |      |              |          |                 0
   11561 | helvetia pl  |               |       |      |              |          |                 0
   44066 |  helvetia    | Helvetia      |       |      |              |          |                 0
   49872 |  helvetia pl | Helvetiaplatz |       |      |              |          |                 0
    1421 | helvetia     |               |       |      |              |          |               158
      16 | pl           |               |       |      |              |          |            435656
```

Now find the place_ids relevant for word 'Helvetiaplatz'

`select place_id, 0::int as exactmatch from search_name where name_vector @> ARRAY[49872] order by (case when importance = 0 OR importance IS NULL then 0.75-(search_rank::float/40) else importance end) DESC limit 20`

```
 place_id | exactmatch 
----------+------------
   158698 |          0
   100153 |          0
   101370 |          0
   103124 |          0
```

**2. with the place_ids found nominatim fetches the display_name labelled as langaddress with this rather long query:**

`select osm_type,osm_id,class,type,admin_level,rank_search,rank_address,min(place_id) as place_id, min(parent_place_id) as parent_place_id, calculated_country_code as country_code,get_address_by_language(place_id, ARRAY['short_name:en-US','short_name:en','short_name:de','name:en-US','name:en','name:de','place_name:en-US','place_name:en','place_name:de','official_name:en-US','official_name:en','official_name:de','short_name','name','place_name','official_name','ref','type']) as langaddress,get_name_by_language(name, ARRAY['short_name:en-US','short_name:en','short_name:de','name:en-US','name:en','name:de','place_name:en-US','place_name:en','place_name:de','official_name:en-US','official_name:en','official_name:de','short_name','name','place_name','official_name','ref','type']) as placename,get_name_by_language(name, ARRAY['ref']) as ref,avg(ST_X(centroid)) as lon,avg(ST_Y(centroid)) as lat, coalesce(importance,0.75-(rank_search::float/40)) as importance, (select max(p.importance*(p.rank_address+2)) from place_addressline s, placex p where s.place_id = min(CASE WHEN placex.rank_search < 28 THEN placex.place_id ELSE placex.parent_place_id END) and p.place_id = s.address_place_id and s.isaddress and p.importance is not null) as addressimportance, (extratags->'place') as extra_place from placex where place_id in (158698,100153,101370,103124) and (placex.rank_address between 0 and 30  OR (extratags->'place') = 'city') and linked_place_id is null group by osm_type,osm_id,class,type,admin_level,rank_search,rank_address,calculated_country_code,importance,langaddress ,placename ,ref ,extratags->'place'  union select 'T' as osm_type,place_id as osm_id,'place' as class,'house' as type,null as admin_level,30 as rank_search,30 as rank_address,min(place_id) as place_id, min(parent_place_id) as parent_place_id,'us' as country_code,get_address_by_language(place_id, ARRAY['short_name:en-US','short_name:en','short_name:de','name:en-US','name:en','name:de','place_name:en-US','place_name:en','place_name:de','official_name:en-US','official_name:en','official_name:de','short_name','name','place_name','official_name','ref','type']) as langaddress,null as placename,null as ref,avg(ST_X(centroid)) as lon,avg(ST_Y(centroid)) as lat, -1.15 as importance, (select max(p.importance*(p.rank_address+2)) from place_addressline s, placex p where s.place_id = min(location_property_tiger.parent_place_id) and p.place_id = s.address_place_id and s.isaddress and p.importance is not null) as addressimportance, null as extra_place from location_property_tiger where place_id in (158698,100153,101370,103124) and 30 between 0 and 30 group by place_id union select 'L' as osm_type,place_id as osm_id,'place' as class,'house' as type,null as admin_level,30 as rank_search,30 as rank_address,min(place_id) as place_id, min(parent_place_id) as parent_place_id,'us' as country_code,get_address_by_language(place_id, ARRAY['short_name:en-US','short_name:en','short_name:de','name:en-US','name:en','name:de','place_name:en-US','place_name:en','place_name:de','official_name:en-US','official_name:en','official_name:de','short_name','name','place_name','official_name','ref','type']) as langaddress,null as placename,null as ref,avg(ST_X(centroid)) as lon,avg(ST_Y(centroid)) as lat, -1.10 as importance, (select max(p.importance*(p.rank_address+2)) from place_addressline s, placex p where s.place_id = min(location_property_aux.parent_place_id) and p.place_id = s.address_place_id and s.isaddress and p.importance is not null) as addressimportance, null as extra_place from location_property_aux where place_id in (158698,100153,101370,103124) and 30 between 0 and 30 group by place_id,get_address_by_language(place_id, ARRAY['short_name:en-US','short_name:en','short_name:de','name:en-US','name:en','name:de','place_name:en-US','place_name:en','place_name:de','official_name:en-US','official_name:en','official_name:de','short_name','name','place_name','official_name','ref','type'])  order by importance desc`

The result looks as follows (look for langaddress):
```
osm_type |  osm_id   |  class  |    type    | admin_level | rank_search | rank_address | place_id | parent_place_id | country_code |                                                  langaddress                                                  |   placename   |      ref      |       lon        |     lat     | importance | addressimportance | extra_place 
W        |   4533221 | highway | pedestrian |          15 |          26 |           26 |   158698 |          422112 | ch           | Helvetiaplatz, Langstrasse, Aussersihl, Zurich, Bezirk Z<C3><BC>rich, Zurich, 8004, Switzerland                     | Helvetiaplatz | Helvetiaplatz | 8.52675220198062 | 47.37617335 |        0.1 |  
                 | 
 N        |  29798259 | railway | tram_stop  |          15 |          30 |           30 |   100153 |          388060 | ch           | Helvetiaplatz, Stauffacherstrasse, Langstrasse, Aussersihl, Zurich, Bezirk Z<C3><BC>rich, Zurich, 8004, Switzerland | Helvetiaplatz | Helvetiaplatz |        8.5260589 |  47.3762266 |          0 |                   | 
 N        | 313898344 | amenity | pharmacy   |          15 |          30 |           30 |   103124 |          323019 | ch           | Helvetiaplatz, Anwandstrasse, Langstrasse, Aussersihl, Zurich, Bezirk Z<C3><BC>rich, Zurich, 8004, Switzerland      | Helvetiaplatz | Helvetiaplatz |         8.525242 |  47.3760616 |          0 |                   | 
 N        | 262153504 | highway | bus_stop   |          15 |          30 |           30 |   101370 |          250636 | ch           | Helvetiaplatz, Langstrasse, Aussersihl, Zurich, Bezirk Z<C3><BC>rich, Zurich, 8004, Switzerland                     | Helvetiaplatz | Helvetiaplatz |        8.5255547 |  47.3762109 |          0 |                   | 
```

Key in this query is once again the table _placex_
see #4 for details
Further, the following tables are relevant:
```
              Table "public.search_name"
       Column       |         Type         | Modifiers 
--------------------+----------------------+-----------
 place_id           | bigint               | 
 search_rank        | integer              | 
 address_rank       | integer              | 
 importance         | double precision     | 
 country_code       | character varying(2) | 
 name_vector        | integer[]            | 
 nameaddress_vector | integer[]            | 
 centroid           | geometry             | 



          Table "public.place_addressline"
       Column        |       Type       | Modifiers 
---------------------+------------------+-----------
 place_id            | bigint           | 
 address_place_id    | bigint           | 
 fromarea            | boolean          | 
 isaddress           | boolean          | 
 distance            | double precision | 
 cached_rank_address | integer          | 

  Table "public.location_property_tiger"
     Column      |   Type   | Modifiers 
-----------------+----------+-----------
 place_id        | bigint   | 
 partition       | integer  | 
 parent_place_id | bigint   | 
 housenumber     | text     | 
 postcode        | text     | 
 centroid        | geometry | 
```

And finally the SQL functions
```
CREATE OR REPLACE FUNCTION get_address_by_language(for_place_id BIGINT, languagepref TEXT[]) RETURNS TEXT
  AS $$
DECLARE
  result TEXT[];
  currresult TEXT;
  prevresult TEXT;
  location RECORD;
BEGIN

  result := '{}';
  prevresult := '';

  FOR location IN select * from get_addressdata(for_place_id) where isaddress order by rank_address desc LOOP
    currresult := trim(get_name_by_language(location.name, languagepref));
    IF currresult != prevresult AND currresult IS NOT NULL AND result[(100 - location.rank_address)] IS NULL THEN
      result[(100 - location.rank_address)] := trim(get_name_by_language(location.name, languagepref));
      prevresult := currresult;
    END IF;
  END LOOP;

  RETURN array_to_string(result,', ');
END;
```

and

```
CREATE OR REPLACE FUNCTION get_addressdata(in_place_id BIGINT) RETURNS setof addressline 
  AS $$
DECLARE
  for_place_id BIGINT;
  result TEXT[];
  search TEXT[];
  found INTEGER;
  location RECORD;
  countrylocation RECORD;
  searchcountrycode varchar(2);
  searchhousenumber TEXT;
  searchhousename HSTORE;
  searchrankaddress INTEGER;
  searchpostcode TEXT;
  searchclass TEXT;
  searchtype TEXT;
  countryname HSTORE;
  hadcountry BOOLEAN;
BEGIN

  select parent_place_id,'us', housenumber, 30, postcode, null, 'place', 'house' from location_property_tiger 
    WHERE place_id = in_place_id 
    INTO for_place_id,searchcountrycode, searchhousenumber, searchrankaddress, searchpostcode, searchhousename, searchclass, searchtype;

  IF for_place_id IS NULL THEN
    select parent_place_id,'us', housenumber, 30, postcode, null, 'place', 'house' from location_property_aux
      WHERE place_id = in_place_id 
      INTO for_place_id,searchcountrycode, searchhousenumber, searchrankaddress, searchpostcode, searchhousename, searchclass, searchtype;
  END IF;

  IF for_place_id IS NULL THEN
    select parent_place_id, calculated_country_code, housenumber, rank_search, postcode, name, class, type from placex 
      WHERE place_id = in_place_id and rank_address = 30 
      INTO for_place_id, searchcountrycode, searchhousenumber, searchrankaddress, searchpostcode, searchhousename, searchclass, searchtype;
  END IF;

  IF for_place_id IS NULL THEN
    select coalesce(linked_place_id, place_id),  calculated_country_code,
           housenumber, rank_search, postcode, null
      from placex where place_id = in_place_id
      INTO for_place_id, searchcountrycode, searchhousenumber, searchrankaddress, searchpostcode, searchhousename;
  END IF;

--RAISE WARNING '% % % %',searchcountrycode, searchhousenumber, searchrankaddress, searchpostcode;

  found := 1000;
  hadcountry := false;
  FOR location IN 
    select placex.place_id, osm_type, osm_id,
      CASE WHEN class = 'place' and type = 'postcode' THEN hstore('name', postcode) ELSE name END as name,
      class, type, admin_level, true as fromarea, true as isaddress,
      CASE WHEN rank_address = 0 THEN 100 WHEN rank_address = 11 THEN 5 ELSE rank_address END as rank_address,
      0 as distance, calculated_country_code, postcode
      from placex
      where place_id = for_place_id 
  LOOP
--RAISE WARNING '%',location;
    IF searchcountrycode IS NULL AND location.calculated_country_code IS NOT NULL THEN
      searchcountrycode := location.calculated_country_code;
    END IF;
    IF searchpostcode IS NOT NULL and location.type = 'postcode' THEN
      location.isaddress := FALSE;
    END IF;
    IF searchpostcode IS NULL and location.postcode IS NOT NULL THEN
      searchpostcode := location.postcode;
    END IF;
    IF location.rank_address = 4 AND location.isaddress THEN
      hadcountry := true;
    END IF;
    IF location.rank_address < 4 AND NOT hadcountry THEN
      select name from country_name where country_code = searchcountrycode limit 1 INTO countryname;
      IF countryname IS NOT NULL THEN
        countrylocation := ROW(null, null, null, countryname, 'place', 'country', null, true, true, 4, 0)::addressline;
        RETURN NEXT countrylocation;
      END IF;
    END IF;
    countrylocation := ROW(location.place_id, location.osm_type, location.osm_id, location.name, location.class, 
                           location.type, location.admin_level, location.fromarea, location.isaddress, location.rank_address, 
                           location.distance)::addressline;
    RETURN NEXT countrylocation;
    found := location.rank_address;
  END LOOP;

  FOR location IN 
    select placex.place_id, osm_type, osm_id,
      CASE WHEN class = 'place' and type = 'postcode' THEN hstore('name', postcode) ELSE name END as name,
      CASE WHEN extratags ? 'place' THEN 'place' ELSE class END as class,
      CASE WHEN extratags ? 'place' THEN extratags->'place' ELSE type END as type,
      admin_level, fromarea, isaddress,
      CASE WHEN address_place_id = for_place_id AND rank_address = 0 THEN 100 WHEN rank_address = 11 THEN 5 ELSE rank_address END as rank_address,
      distance,calculated_country_code,postcode
      from place_addressline join placex on (address_place_id = placex.place_id) 
      where place_addressline.place_id = for_place_id 
      and (cached_rank_address > 0 AND cached_rank_address < searchrankaddress)
      and address_place_id != for_place_id
      and (placex.calculated_country_code IS NULL OR searchcountrycode IS NULL OR placex.calculated_country_code = searchcountrycode)
      order by rank_address desc,isaddress desc,fromarea desc,distance asc,rank_search desc
  LOOP
--RAISE WARNING '%',location;
    IF searchcountrycode IS NULL AND location.calculated_country_code IS NOT NULL THEN
      searchcountrycode := location.calculated_country_code;
    END IF;
    IF searchpostcode IS NOT NULL and location.type = 'postcode' THEN
      location.isaddress := FALSE;
    END IF;
    IF searchpostcode IS NULL and location.isaddress and location.type != 'postcode' and location.postcode IS NOT NULL THEN
      searchpostcode := location.postcode;
    END IF;
    IF location.rank_address = 4 AND location.isaddress THEN
      hadcountry := true;
    END IF;
    IF location.rank_address < 4 AND NOT hadcountry THEN
      select name from country_name where country_code = searchcountrycode limit 1 INTO countryname;
      IF countryname IS NOT NULL THEN
        countrylocation := ROW(null, null, null, countryname, 'place', 'country', null, true, true, 4, 0)::addressline;
        RETURN NEXT countrylocation;
      END IF;
    END IF;
    countrylocation := ROW(location.place_id, location.osm_type, location.osm_id, location.name, location.class, 
                           location.type, location.admin_level, location.fromarea, location.isaddress, location.rank_address, 
                           location.distance)::addressline;
    RETURN NEXT countrylocation;
    found := location.rank_address;
  END LOOP;

  IF found > 4 THEN
    select name from country_name where country_code = searchcountrycode limit 1 INTO countryname;
--RAISE WARNING '% % %',found,searchcountrycode,countryname;
    IF countryname IS NOT NULL THEN
      location := ROW(null, null, null, countryname, 'place', 'country', null, true, true, 4, 0)::addressline;
      RETURN NEXT location;
    END IF;
  END IF;

  IF searchcountrycode IS NOT NULL THEN
    location := ROW(null, null, null, hstore('ref', searchcountrycode), 'place', 'country_code', null, true, false, 4, 0)::addressline;
    RETURN NEXT location;
  END IF;

  IF searchhousename IS NOT NULL THEN
    location := ROW(in_place_id, null, null, searchhousename, searchclass, searchtype, null, true, true, 29, 0)::addressline;
--    location := ROW(in_place_id, null, null, searchhousename, 'place', 'house_name', null, true, true, 29, 0)::addressline;
    RETURN NEXT location;
  END IF;

  IF searchhousenumber IS NOT NULL THEN
    location := ROW(in_place_id, null, null, hstore('ref', searchhousenumber), 'place', 'house_number', null, true, true, 28, 0)::addressline;
    RETURN NEXT location;
  END IF;

  IF searchpostcode IS NOT NULL THEN
    location := ROW(null, null, null, hstore('ref', searchpostcode), 'place', 'postcode', null, true, true, 5, 0)::addressline;
    RETURN NEXT location;
  END IF;

  RETURN;
END;
```

**Comment by and01** [comment]
**How to Nominatim links geometry to place nodes (Dorf/Gemeinde)?**

- the geometry (often times a polygon) is stored in table placex as geometry object
- the node of the place (in nomiatim details page called 'centre point') is stored as centroid in table placex also as a geometry object

Therefore, this information is derived during the import and not on query time

**Comment by and01** [comment]
**Hierarchy (belongs to Switzerland, belongs to Kanton, etc) reconstruction**

example
https://nominatim.openstreetmap.org/details.php?place_id=145098706 (ids are different in local nominatim installation)

The following SQL query fetches the data for the 'parent of section' in the details page:

`select obj.place_id, osm_type, osm_id, class, type, housenumber, admin_level, rank_address, ST_GeometryType(geometry) in ('ST_Polygon','ST_MultiPolygon') as isarea, st_distance(geometry, placegeometry) as distance,  get_name_by_language(name,ARRAY['short_name:en-US','short_name:en','short_name:de','name:en-US','name:en','name:de','place_name:en-US','place_name:en','place_name:de','official_name:en-US','official_name:en','official_name:de','short_name','name','place_name','official_name','ref','type']) as localname, length(name::text) as namelength  from (select placex.place_id, osm_type, osm_id, class, type, housenumber, admin_level, rank_address, rank_search, geometry, name from placex  where parent_place_id = 422106 order by rank_address asc,rank_search asc limit 500) as obj, (select geometry as placegeometry from placex where place_id = 422106) as x order by rank_address asc,rank_search asc,localname,housenumber`

The result in this example is as an the details page:
```
 place_id | osm_type |   osm_id   |  class   |      type      | housenumber | admin_level | rank_address | isarea | distance |  localname   | namelength 
----------+----------+------------+----------+----------------+-------------+-------------+--------------+--------+----------+--------------+------------
   100533 | N        |   98161723 | place    | locality       |             |          15 |            0 | f      |        0 | Brunaupark   |         20
   128842 | N        | 2012632412 | place    | locality       |             |          15 |            0 | f      |        0 | Goldbrunnegg |         22
   422107 | R        |    1694107 | boundary | administrative |             |          10 |           20 | t      |        0 | Alt-Wiedikon |         22
   422108 | R        |    1694108 | boundary | administrative |             |          10 |           20 | t      |        0 | Friesenberg  |         21
   422109 | R        |    1694109 | boundary | administrative |             |          10 |           20 | t      |        0 | Sihlfeld     |         18
```

The following SQL query fetches the data for the 'address section' in the details page:

`select *,get_name_by_language(name,ARRAY['short_name:en-US','short_name:en','short_name:de','name:en-US','name:en','name:de','place_name:en-US','place_name:en','place_name:de','official_name:en-US','official_name:en','official_name:de','short_name','name','place_name','official_name','ref','type']) as localname from get_addressdata(422106) order by rank_address desc,isaddress desc;
`

Just to verifiy I wrote the following to get the hierarchy for "Wiedikon" which has place_id = 422106

`select rank_address, isaddress, name->'name' as name from get_addressdata(422106) order by rank_address desc,isaddress desc;`

The result is what you are looking for:
```
 rank_address | isaddress |      name      
--------------+-----------+----------------
           18 | t         | Wiedikon
           16 | t         | Zürich
           12 | t         | Bezirk Zürich
            8 | t         | Zürich
            4 | t         | Switzerland
            4 | f         | 
```

see function get_addressdata above for further details
see #4 for creating the rank_address

In a nutshell, it basically checks for parent_place_ids until no parent_id is set. Manually, it could looks like this:
```

nominatim=# SELECT place_id, parent_place_id, isin,  name->'name' FROM placex p WHERE p.place_id = 422106;
 place_id | parent_place_id | isin | ?column? 
----------+-----------------+------+----------
   422106 |          422035 |      | Wiedikon

 nominatim=# SELECT place_id, parent_place_id, isin,  name->'name' FROM placex p WHERE p.place_id = 422035;
 place_id | parent_place_id | isin  | ?column? 
----------+-----------------+-------+----------
   422035 |          422094 | CH-ZH | Zürich
 
 nominatim=# SELECT place_id, parent_place_id, isin,  name->'name' FROM placex p WHERE p.place_id = 422094;
 place_id | parent_place_id | isin |    ?column?    
----------+-----------------+------+----------------
   422094 |          103123 |      | Bezirk Zürich

nominatim=# SELECT place_id, parent_place_id, isin,  name->'name' FROM placex p WHERE p.place_id = 103123;
 place_id | parent_place_id |        isin        | ?column? 
----------+-----------------+--------------------+----------
   103123 |               0 | Europe,Switzerland | Zürich   
```

**Comment by klokan** [comment]
Thank you a lot @and01!

Until now the notes here are about query time processing in Nominatim - the indexing time hierarchy reconstraction analyse is in #5
